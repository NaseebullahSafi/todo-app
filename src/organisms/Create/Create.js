import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { v4 as uuidv4 } from "uuid";

import { addTodo } from "src/Store/todos/actions";
import { addCreateToStack } from "src/Store/userActions/actions";
import Input from "src/atoms/Input";
import TextArea from "src/atoms/TextArea";
import Button from "src/atoms/Button";
import {
  StyledForm,
  StyledInputContainer,
  StyledTextAreaContainer,
} from "./Create.style";

const Create = ({ onSubmit }) => {
  const { userActions } = useSelector(({ userActions }) => ({ userActions }));
  const dispatch = useDispatch();
  const [id, setId] = React.useState("");
  const [desc, setDesc] = React.useState("");
  const [date, setDate] = React.useState("");
  const [name, setName] = React.useState("");

  const handleChange = (e) => {
    switch (e.target.name) {
      case "name":
        setName(e.target.value);
        return;
      case "description":
        setDesc(e.target.value);
        return;
    }
  };

  /**
   * Basic form validation to ensure all the fields are populated.
   * TODO: improve on the validation of the inputs with regex and
   * string rules.
   */
  const validateForm = () => {
    if (desc === "" || name === "") return false;
    return true;
  };

  const clearForm = () => {
    setId("");
    setDesc("");
    setDate("");
    setName("");
  };

  /**
   * Will submit the form. i.e. if successful, store the values to localStorage
   * @param {*} e
   */
  const onSubmitForm = (e) => {
    e.preventDefault();

    const data = {
      id: uuidv4(),
      name,
      description: desc,
      date: new Date().toString(),
    };

    // is it a controlled component? This checks that.
    if (typeof onSubmit === "function") {
      onSubmit(data);
      return;
    }

    // form is validated
    if (validateForm()) {
      // dispatch the action which will add the todo to redux
      dispatch(addTodo(data));

      // if its recording, then push to our action stack in redux
      if (userActions.isRecording) {
        dispatch(
          addCreateToStack({
            id: data.id,
            action: "ADD",
          })
        );
      }
    }
    clearForm();
  };

  return (
    <StyledForm onSubmit={onSubmitForm}>
      <h1>Create new Todo task</h1>
      <StyledInputContainer>
        <label htmlFor="name">Name</label>
        <Input
          type="text"
          name="name"
          id="name"
          value={name}
          onChange={handleChange}
        />
      </StyledInputContainer>
      <StyledTextAreaContainer>
        <label htmlFor="description">Description</label>
        <TextArea
          type="text"
          name="description"
          id="description"
          value={desc}
          onChange={handleChange}
        />
      </StyledTextAreaContainer>

      <Button type="submit" value="submit">
        Submit
      </Button>
    </StyledForm>
  );
};

export default Create;
