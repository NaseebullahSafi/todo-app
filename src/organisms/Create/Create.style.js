import styled from "styled-components";

export const StyledForm = styled.form`
  padding: 10px;
  display: flex;
  flex-direction: column;
  align-items: stretch;
  justify-content: center;
`;

export const StyledInputContainer = styled.div``;

export const StyledTextAreaContainer = styled.div`
  margin-top: 3em;
  min-height: 100px;
`;
