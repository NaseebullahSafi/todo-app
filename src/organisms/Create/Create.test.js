import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import "jest-styled-components";
import { v4 as uuidv4 } from "uuid";

import Create from "./Create";

describe("<Create />", () => {
  let wrapper;

  const setState = jest.fn();
  const onSubmitSpy = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");

  useStateSpy.mockImplementation((init) => [init, setState]);

  beforeEach(() => {
    wrapper = mount(shallow(<Create onSubmit={onSubmitSpy} />).get(0));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render Create", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Name input", () => {
    it("should capture name correctly onChange", () => {
      const name = wrapper.find("#name").at(0);
      name.instance().value = "Test";
      name.simulate("change");
      expect(setState).toHaveBeenCalledWith("Test");
    });
  });

  describe("Description input", () => {
    it("should capture description correctly onChange", () => {
      const description = wrapper.find("#description").at(0);
      description.instance().value = "Lorem ipsum";
      description.simulate("change");
      expect(setState).toHaveBeenCalledWith("Lorem ipsum");
    });
  });

  describe("Form submit", () => {
    it("should submit form successfully", () => {
      wrapper.find("form").simulate("submit", { preventDefault: () => {} });
      expect(onSubmitSpy).toHaveBeenCalledTimes(1);
    });
  });
});
