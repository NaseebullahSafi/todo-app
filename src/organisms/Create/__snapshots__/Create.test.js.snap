// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Create /> should render Create 1`] = `"<form class=\\"Createstyle__StyledForm-sc-9trsr-0 ijeoyf\\"><h1>Create new Todo task</h1><div class=\\"Createstyle__StyledFormInput-sc-9trsr-1 FvLXJ\\"><label for=\\"name\\">Name</label><input type=\\"text\\" name=\\"name\\" id=\\"name\\" class=\\"Inputstyle__StyledInput-sc-1f6guk6-0 bewcis\\" value=\\"\\"></div><div class=\\"Createstyle__StyledFormTextArea-sc-9trsr-2 bJuplU\\"><label for=\\"description\\">Description</label><textarea type=\\"text\\" name=\\"description\\" id=\\"description\\" class=\\"TextAreastyle__StyledTextArea-sc-1dz7fyc-0 jvRtvK\\"></textarea></div><button type=\\"submit\\" value=\\"submit\\" class=\\"Buttonstyle__StyledButton-sc-1atuls6-0 ejDqxw\\">Create</button></form>"`;
